### PR TITLE
Implement `determinant` f32 tests

### PR DESCRIPTION
**Pull request by zoddicus**
Issue #1245

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [x] Tests are properly located in the test tree.
- [x] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [x] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [x] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/unittests/f32_interval.spec.ts [hunk 1]
```diff
@@ -3425,6 +3426,122 @@ g.test('modfInterval')
...
+      ],
+      expected: -3,
+    },
+    {
+      input: [
+        [1, 2, 3],
+        [4, 5, 6],
+        [7, 8, 9],
+      ],
+      expected: 0,
```
Comment (dneto0) [inline src/unittests/f32_interval.spec.ts:3489]: Surprising, but correct.

I found an online tool for this from Wolfram Alpha.
https://www.wolframalpha.com/input?i2d=true&i=determinant%7B%7B1%2C2%2C3%7D%2C%7B4%2C5%2C6%7D%2C%7B7%2C8%2C9%7D%7D

File: src/webgpu/util/math.ts [hunk 2]
```diff
@@ -1128,6 +1128,14 @@ export function cartesianProduct<T>(...inputs: T[][]): T[][] {
  * Recursively calculates all of the permutations, does not cull duplicate
  * entries.
  *
+ * Only feasible for inputs of lengths 5 or so, since the number of permutations
+ * is (inpyt.length)!, so will cause the stack to explode for longer inputs.
+ *
+ * This code could be made iterative using something like
+ * Steinhaus–Johnson–Trotter and additionally turned into a generator to reduce
```
Comment (dneto0) [inline src/webgpu/util/math.ts:1135]: TIL!

File: src/webgpu/util/f32_interval.ts [hunk 3]
```diff
@@ -1638,6 +1654,148 @@ export function degreesInterval(n: number): F32Interval {
...
+  // For simplicity sake A, B, C are calculated as the elements of the first
+  // column
+  const A = multiplicationInterval(m[0][0], determinant2x2Interval(minorNxN(m, 0, 0)));
+  const B = multiplicationInterval(
+    m[0][1],
+    multiplicationInterval(-1, determinant2x2Interval(minorNxN(m, 0, 1)))
+  );
+  const C = multiplicationInterval(m[0][2], determinant2x2Interval(minorNxN(m, 0, 2)));
+
+  // Need to calculate permutations, since for fp addition is not transitive,
```
Comment (dneto0) [inline src/webgpu/util/f32_interval.ts:1715]: nit:  s/transitive/associative/

It's the associativity that's the issue.

  (a+b)+c  is not always a + (b+c).

Cummutativity is   a + b = b + c, and that does hold for float math.

File: src/webgpu/util/f32_interval.ts [hunk 4]
```diff
@@ -1638,6 +1654,148 @@ export function degreesInterval(n: number): F32Interval {
...
+  // det(M) is A + B + C + D, where A, B, C, D are four elements in a row/column
+  // times their own co-factor.
+  // (The co-factor is the determinant of the minor of that position with the
+  // appropriate +/-)
+  // For simplicity sake A, B, C, D are calculated as the elements of the
+  // first column
+  const A = multiplicationInterval(m[0][0], determinant3x3Interval(minorNxN(m, 0, 0)));
+  const B = multiplicationInterval(
+    m[0][1],
+    multiplicationInterval(-1, determinant3x3Interval(minorNxN(m, 0, 1)))
```
Comment (dneto0) [inline src/webgpu/util/f32_interval.ts:1742]: Can use -m[0][1] here too, and once more below.

**Review by dneto0 — APPROVED** [review]
Very nice.  Only nits.

Thanks!

File: src/webgpu/util/math.ts (hunk 2 above)
Comment (dneto0) [reply src/webgpu/util/math.ts:1135]: Also, this comment led me to "change ringing" on wikipedia, and then on my music streaming service. So thanks.  :-)

Comment (zoddicus) [reply src/webgpu/util/math.ts:1135]: A comment so powerful it nerd snipes the reviewer :+1:

**Pull request merged**